File: windows/src/BabyMonitor.Core/Json/Json.cs
using System.Globalization;
using System.Text;
using System.Text.Json;

namespace BabyMonitor.Core.Json;

/// <summary>
/// A small, permissive JSON object model — the C# twin of core's <c>json/Json.kt</c>.
///
/// The protocol layer reads Xiaomi's untyped, inconsistent responses through lenient <c>Opt*</c>
/// accessors: a missing or wrong-typed field reads as its default and never throws. That lenience is
/// the point — Xiaomi's responses vary by region, account state and firmware, and a monitor that
/// dies on an unexpected field at 3am is worse than one that carries on with a default.
///
/// Keeping the same shape as the Kotlin model is deliberate: the port stays line-for-line
/// comparable with the implementation that is known to talk to the real gateway.
///
/// Values are held as string, long, double, bool, JsonObj, JsonArr or null.
/// </summary>
public sealed class JsonObj
{
    private readonly Dictionary<string, object?> _values;
    private readonly List<string> _order; // insertion order: the signed form depends on it

    public JsonObj()
    {
        _values = new Dictionary<string, object?>(StringComparer.Ordinal);
        _order = new List<string>();
    }

    internal JsonObj(Dictionary<string, object?> values, List<string> order)
    {
        _values = values;
        _order = order;
    }

    public JsonObj(string json)
        : this()
    {
        JsonElement root;
        try
        {
            using var doc = JsonDocument.Parse(json, new JsonDocumentOptions
            {
                AllowTrailingCommas = true,
                CommentHandling = JsonCommentHandling.Skip,
            });
            root = doc.RootElement.Clone();
        }
        catch (Exception e)
        {
            throw new JsonException($"not JSON: {e.Message}");
        }

        if (root.ValueKind != JsonValueKind.Object)
        {
            throw new JsonException("not a JSON object");
        }

        foreach (var property in root.EnumerateObject())
        {
            Put(property.Name, JsonConvert.FromElement(property.Value));
        }
    }

    public JsonObj Put(string key, object? value)
    {
        if (!_values.ContainsKey(key))
        {
            _order.Add(key);
        }

        _values[key] = JsonConvert.Normalize(value);
        return this;
    }

    public bool Has(string key) => _values.ContainsKey(key);

    public bool IsNull(string key) => !_values.TryGetValue(key, out var v) || v is null;

    public object? Opt(string key) => _values.GetValueOrDefault(key);

    public object Get(string key) =>
        _values.GetValueOrDefault(key) ?? throw new JsonException($"no value for '{key}'");

    public string GetString(string key) => JsonConvert.AsString(Get(key));

    /// <summary>Containers have no string form worth having — they read as the default.</summary>
    public string OptString(string key, string fallback = "")
    {
        var v = _values.GetValueOrDefault(key);
        return v switch
        {
            null => fallback,
            JsonObj => fallback,
            JsonArr => fallback,
            _ => JsonConvert.AsString(v),
        };
    }

    public int OptInt(string key, int fallback = 0)
    {
        var n = JsonConvert.AsNumber(_values.GetValueOrDefault(key));
        return n is null ? fallback : (int)n.Value;
    }

    public double OptDouble(string key, double fallback) =>
        JsonConvert.AsNumber(_values.GetValueOrDefault(key)) ?? fallback;

    public bool OptBoolean(string key, bool fallback = false) =>
        _values.GetValueOrDefault(key) switch
        {
            bool b => b,
            string s => bool.TryParse(s, out var parsed) ? parsed : fallback,
            _ => fallback,
        };

    public int GetInt(string key)
    {
        var n = JsonConvert.AsNumber(Get(key));
        return n is null ? throw new JsonException($"'{key}' is not a number") : (int)n.Value;
    }

    public JsonObj GetJsonObject(string key) =>
        _values.GetValueOrDefault(key) as JsonObj ?? throw new JsonException($"'{key}' is not an object");

    public JsonObj? OptJsonObject(string key) => _values.GetValueOrDefault(key) as JsonObj;

    public JsonArr? OptJsonArray(string key) => _values.GetValueOrDefault(key) as JsonArr;

    public override string ToString()
    {
        var out_ = new StringBuilder();
        JsonConvert.WriteObject(out_, _order, _values);
        return out_.ToString();
    }
}

public sealed class JsonArr
{
    private readonly List<object?> _values;

    public JsonArr() => _values = new List<object?>();

    internal JsonArr(List<object?> values) => _values = values;

    public JsonArr Put(object? value)
    {
        _values.Add(JsonConvert.Normalize(value));
        return this;
    }

    public int Length => _values.Count;

    public object? Opt(int index) => index >= 0 && index < _values.Count ? _values[index] : null;

    public JsonObj GetJsonObject(int index) =>
        Opt(index) as JsonObj ?? throw new JsonException($"[{index}] is not an object");

    public JsonObj? OptJsonObject(int index) => Opt(index) as JsonObj;

    public override string ToString()
    {
        var out_ = new StringBuilder();
        JsonConvert.WriteArray(out_, _values);
        return out_.ToString();
    }
}

public sealed class JsonException : Exception
{
    public JsonException(string message)
        : base(message)
    {
    }
}

internal static class JsonConvert
{
    internal static object? FromElement(JsonElement element) => element.ValueKind switch
    {
        JsonValueKind.Null or JsonValueKind.Undefined => null,
        JsonValueKind.True => true,
        JsonValueKind.False => false,
        JsonValueKind.String => element.GetString(),
        // The cast is load-bearing: without it the conditional's type is `double`, and every whole
        // number in a Xiaomi response would arrive as 2.0 rather than 2 — which is not what the other
        // implementations hold, and is not what a MiOT property value compares equal to.
        JsonValueKind.Number => element.TryGetInt64(out var l) ? (object)l : element.GetDouble(),
        JsonValueKind.Object => ObjectFrom(element),
        JsonValueKind.Array => ArrayFrom(element),
        _ => null,
    };

    private static JsonObj ObjectFrom(JsonElement element)
    {
        var obj = new JsonObj();
        foreach (var property in element.EnumerateObject())
        {
            obj.Put(property.Name, FromElement(property.Value));
        }

        return obj;
    }

    private static JsonArr ArrayFrom(JsonElement element)
    {
        var arr = new JsonArr();
        foreach (var item in element.EnumerateArray())
        {
            arr.Put(FromElement(item));
        }

        return arr;
    }

    internal static object? Normalize(object? value) => value switch
    {
        int i => (long)i,
        short s => (long)s,
        byte b => (long)b,
        float f => (double)f,
        _ => value,
    };

    internal static string AsString(object value) => value switch
    {
        string s => s,
        double d => FormatDouble(d),
        bool b => b ? "true" : "false",
        long l => l.ToString(CultureInfo.InvariantCulture),
        _ => value.ToString() ?? string.Empty,
    };

    internal static double? AsNumber(object? value) => value switch
    {
        long l => l,
        double d => d,
        bool b => b ? 1.0 : 0.0,
        string s => double.TryParse(s, NumberStyles.Float, CultureInfo.InvariantCulture, out var d) ? d : null,
        _ => null,
    };

    /// <summary>A whole-number double reads back as a whole number, not "5.0".</summary>
    internal static string FormatDouble(double value) =>
        double.IsFinite(value) && value == Math.Truncate(value) && Math.Abs(value) < 9e18
            ? ((long)value).ToString(CultureInfo.InvariantCulture)
            : value.ToString("R", CultureInfo.InvariantCulture);

    internal static void WriteObject(StringBuilder out_, List<string> order, Dictionary<string, object?> values)
    {
        out_.Append('{');
        var first = true;
        foreach (var key in order)
        {
            if (!first)
            {
                out_.Append(',');
            }

            first = false;
            WriteString(out_, key);
            out_.Append(':');
            WriteValue(out_, values.GetValueOrDefault(key));
        }

        out_.Append('}');
    }

    internal static void WriteArray(StringBuilder out_, List<object?> values)
    {
        out_.Append('[');
        for (var i = 0; i < values.Count; i++)
        {
            if (i > 0)
            {
                out_.Append(',');
            }

            WriteValue(out_, values[i]);
        }

        out_.Append(']');
    }

    private static void WriteValue(StringBuilder out_, object? value)
    {
        switch (value)
        {
            case null:
                out_.Append("null");
                break;
            case string s:
                WriteString(out_, s);
                break;
            case bool b:
                out_.Append(b ? "true" : "false");
                break;
            case long l:
                out_.Append(l.ToString(CultureInfo.InvariantCulture));
                break;
            case double d:
                out_.Append(FormatDouble(d));
                break;
            case JsonObj or JsonArr:
                out_.Append(value.ToString());
                break;
            default:
                WriteString(out_, value.ToString() ?? string.Empty);
                break;
        }
    }

    private static void WriteString(StringBuilder out_, string s)
    {
        out_.Append('"');
        foreach (var c in s)
        {
            switch (c)
            {
                case '"':
                    out_.Append("\\\"");
                    break;
                case '\\':
                    out_.Append("\\\\");
                    break;
                case '\n':
                    out_.Append("\\n");
                    break;
                case '\r':
                    out_.Append("\\r");
                    break;
                case '\t':
                    out_.Append("\\t");
                    break;
                default:
                    if (c < ' ')
                    {
                        out_.Append("\\u").Append(((int)c).ToString("x4", CultureInfo.InvariantCulture));
                    }
                    else
                    {
                        out_.Append(c);
                    }

                    break;
            }
        }

        out_.Append('"');
    }
}

File: core/src/commonMain/kotlin/com/bluzi/babymonitor/json/Json.kt
package com.bluzi.babymonitor.json

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray as KxArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject as KxObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * A small, permissive JSON object model.
 *
 * Android ships `org.json` in the platform; nothing else does. Rather than rewrite the protocol
 * layer around a typed serializer — which would mean re-deriving from scratch exactly how the Mi
 * gateway's untyped, inconsistent responses are read — this keeps the same shape and the same
 * lenient `opt*` semantics, with a real parser (kotlinx.serialization) underneath.
 *
 * The lenience is the point: a missing or wrong-typed field reads as its default and never throws.
 * Xiaomi's responses vary by region, account state and firmware, and a monitor that dies on an
 * unexpected field at 3am is worse than one that carries on with a default.
 *
 * Values are held as String, Long, Double, Boolean, JSONObject, JSONArray or null.
 */
class JSONObject internal constructor(private val values: MutableMap<String, Any?>) {
    constructor() : this(mutableMapOf())

    constructor(json: String) : this(parseToObject(json))

    fun put(key: String, value: Any?): JSONObject {
        values[key] = normalize(value)
        return this
    }

    fun has(key: String): Boolean = values.containsKey(key)

    fun isNull(key: String): Boolean = values[key] == null

    fun opt(key: String): Any? = values[key]

    fun get(key: String): Any = values[key] ?: throw JsonException("no value for '$key'")

    fun getString(key: String): String = asString(get(key))

    /** Containers have no string form worth having — they read as the default, as in org.json. */
    fun optString(key: String, default: String = ""): String = when (val v = values[key]) {
        null, is JSONObject, is JSONArray -> default
        else -> asString(v)
    }

    fun optInt(key: String, default: Int = 0): Int = asNumber(values[key])?.toInt() ?: default

    fun optDouble(key: String, default: Double): Double = asNumber(values[key]) ?: default

    fun optBoolean(key: String, default: Boolean = false): Boolean = when (val v = values[key]) {
        is Boolean -> v
        is String -> v.toBooleanStrictOrNull() ?: default
        else -> default
    }

    fun getInt(key: String): Int =
        asNumber(get(key))?.toInt() ?: throw JsonException("'$key' is not a number")

    fun getJSONObject(key: String): JSONObject =
        values[key] as? JSONObject ?: throw JsonException("'$key' is not an object")

    fun getJSONArray(key: String): JSONArray =
        values[key] as? JSONArray ?: throw JsonException("'$key' is not an array")

    fun optJSONObject(key: String): JSONObject? = values[key] as? JSONObject

    fun optJSONArray(key: String): JSONArray? = values[key] as? JSONArray

    override fun toString(): String = buildString { writeObject(this, values) }
}

class JSONArray internal constructor(private val values: MutableList<Any?>) {
    constructor() : this(mutableListOf())

    fun put(value: Any?): JSONArray {
        values.add(normalize(value))
        return this
    }

    fun length(): Int = values.size

    fun opt(index: Int): Any? = values.getOrNull(index)

    fun getJSONObject(index: Int): JSONObject =
        values.getOrNull(index) as? JSONObject ?: throw JsonException("[$index] is not an object")

    fun optJSONObject(index: Int): JSONObject? = values.getOrNull(index) as? JSONObject

    override fun toString(): String = buildString { writeArray(this, values) }
}

class JsonException(message: String) : Exception(message)

// --- parsing ----------------------------------------------------------------

private val parser = Json { isLenient = true; ignoreUnknownKeys = true }

private fun parseToObject(json: String): MutableMap<String, Any?> {
    val element = try {
        parser.parseToJsonElement(json)
    } catch (e: Exception) {
        throw JsonException("not JSON: ${e.message}")
    }
    if (element !is KxObject) throw JsonException("not a JSON object")
    return element.mapValuesTo(mutableMapOf()) { (_, v) -> fromElement(v) }
}

private fun fromElement(element: JsonElement): Any? = when (element) {
    is JsonNull -> null
    is JsonPrimitive ->
        if (element.isString) {
            element.content
        } else {
            element.content.toLongOrNull()
                ?: element.content.toDoubleOrNull()
                ?: element.content.toBooleanStrictOrNull()
                ?: element.content
        }
    is KxObject -> JSONObject(element.mapValuesTo(mutableMapOf()) { (_, v) -> fromElement(v) })
    is KxArray -> JSONArray(element.mapTo(mutableListOf()) { fromElement(it) })
}

// --- coercion ---------------------------------------------------------------

private fun normalize(value: Any?): Any? = when (value) {
    is Int -> value.toLong()
    is Short -> value.toLong()
    is Byte -> value.toLong()
    is Float -> value.toDouble()
    else -> value
}

private fun asString(value: Any): String = when (value) {
    is String -> value
    is Double -> formatDouble(value)
    else -> value.toString()
}

private fun asNumber(value: Any?): Double? = when (value) {
    is Long -> value.toDouble()
    is Double -> value
    is Boolean -> if (value) 1.0 else 0.0
    is String -> value.toDoubleOrNull()
    else -> null
}

/** A whole-number double reads back as a whole number, not "5.0". */
private fun formatDouble(value: Double): String =
    if (value.isFinite() && value == value.toLong().toDouble()) value.toLong().toString()
    else value.toString()

// --- writing ----------------------------------------------------------------

private fun writeObject(out: StringBuilder, values: Map<String, Any?>) {
    out.append('{')
    var first = true
    for ((k, v) in values) {
        if (!first) out.append(',')
        first = false
        writeString(out, k)
        out.append(':')
        writeValue(out, v)
    }
    out.append('}')
}

private fun writeArray(out: StringBuilder, values: List<Any?>) {
    out.append('[')
    for ((i, v) in values.withIndex()) {
        if (i > 0) out.append(',')
        writeValue(out, v)
    }
    out.append(']')
}

private fun writeValue(out: StringBuilder, value: Any?) {
    when (value) {
        null -> out.append("null")
        is String -> writeString(out, value)
        is Boolean, is Long -> out.append(value.toString())
        is Double -> out.append(formatDouble(value))
        is JSONObject, is JSONArray -> out.append(value.toString())
        else -> writeString(out, value.toString())
    }
}

private fun writeString(out: StringBuilder, s: String) {
    out.append('"')
    for (c in s) {
        when {
            c == '"' -> out.append("\\\"")
            c == '\\' -> out.append("\\\\")
            c == '\n' -> out.append("\\n")
            c == '\r' -> out.append("\\r")
            c == '\t' -> out.append("\\t")
            c < ' ' -> out.append("\\u").append(c.code.toString(16).padStart(4, '0'))
            else -> out.append(c)
        }
    }
    out.append('"')
}
